README.md
# @asset-projects/token-list

[![npm](https://img.shields.io/npm/v/@asset-projects/token-list)](https://unpkg.com/@asset-projects/token-list@latest/)[![Package size](https://badgen.net/bundlephobia/minzip/@asset-projects/token-list)](https://bundlephobia.com/package/@asset-projects/token-list)[![!License](https://badgen.net/npm/license/@asset-projects/token-balance)](https://github.com/asset-projects/token-balance/blob/main/LICENSE)

[![Check Markdown links](https://github.com/asset-projects/token-list/actions/workflows/linkchecker.yml/badge.svg)](https://github.com/asset-projects/token-list/actions/workflows/linkchecker.yml)
[![Tests](https://github.com/asset-projects/token-list/actions/workflows/tests.yml/badge.svg)](https://github.com/asset-projects/token-list/actions/workflows/tests.yml)
[![release-please](https://github.com/asset-projects/token-list/actions/workflows/release-please.yml/badge.svg)](https://github.com/asset-projects/token-list/actions/workflows/release-please.yml)

This library maintains a list of ERC20 tokens that exist on the ethereum mainnet, L2 and sidechain.

It manages the main tokens and the top 100 or so tokens.

Support for Ethereum mainnet, optimistic-ethereum, arbitrum, polygon, and Other chains will be supported in the future.

## Install on your project

```zsh
  # npm
  npm i @asset-projects/token-list

  # yarn
  yarn add @asset-projects/token-list
```

# How to Use

```ts
import { getTokenList } from '@asset-projects/token-list';

console.log(getTokenList(1).ERC20);
// [{"chainId": 1, "type": "ERC20", "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599", "name": "Wrapped BTC", ...}]

console.log(getTokenList(1).ERC721);
// [{"chainId": 1, "type": "ERC721", "address": "0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85", "name": "Ethereum Name Service", ...}]
```

# Other imports

```ts
import { tokens } from '@asset-projects/token-list';

console.log(tokens().filter((token) => token.chainId === 1 && token.symbol === 'DAI'));
// [
//   {
//     chainId: 1,
//     type: 'ERC20',
//     address: '0x6b175474e89094c44da98b954eedeac495271d0f',
//     name: 'Dai Stablecoin',
//     symbol: 'DAI',
//     decimals: 18,
//     logoURI: 'https://raw.githubusercontent.com/asset-projects/token-list/main/public/dai.png'
//   }
// ]

console.log(
  tokens().filter(
    (token) =>
      token.chainId === 1 && token.address === '0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85', // ENS
  ),
);
// [
//   {
//     chainId: 1,
//     type: 'ERC721',
//     address: '0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85',
//     name: 'Ethereum Name Service',
//     symbol: 'ENS',
//     logoURI: ''
//   }
// ]
```

## License

MIT


CHANGELOG.md
# Changelog

## [4.2.2](https://github.com/asset-projects/token-list/compare/v4.2.1...v4.2.2) (2022-06-07)


### Bug Fixes

* types ([89ee895](https://github.com/asset-projects/token-list/commit/89ee89596f451ebae3f0ec42f21fe3275c8cc158))

## [4.2.1](https://github.com/asset-projects/token-list/compare/v4.2.0...v4.2.1) (2022-06-05)


### Bug Fixes

* linkchecker.yml ([9af2d36](https://github.com/asset-projects/token-list/commit/9af2d36a023657f16a97fd87bfcfaa27a7e25dfc))
* linkchecker.yml ([9944975](https://github.com/asset-projects/token-list/commit/994497537c989316a17d8f93cda70e6678728fab))

## [4.2.0](https://github.com/asset-projects/token-list/compare/v4...v4.2.0) (2022-06-02)

### Features

- Add Semantic-release to this workflow ([#3](https://github.com/asset-projects/token-list/issues/3)) ([32691c7](https://github.com/asset-projects/token-list/commit/32691c7af5d42ac29065c0d9c425873c86b5eba9))

### Bug Fixes

- ./README.md ([#9](https://github.com/asset-projects/token-list/issues/9)) ([9cec670](https://github.com/asset-projects/token-list/commit/9cec670b8c8029e828fd152278bbf655a7688f48))
- .github/workflows/release.yml ([#5](https://github.com/asset-projects/token-list/issues/5)) ([595c412](https://github.com/asset-projects/token-list/commit/595c4127be5086b98ce5755c3bdca65c8f5b2bd7))
- fixed Github Action ([dd7c8e4](https://github.com/asset-projects/token-list/commit/dd7c8e48112edf1ddc61f7ab076ff8c2cb7ae6e3))
- fixes ([9213c52](https://github.com/asset-projects/token-list/commit/9213c524ea22ed4f2340d536e8488a174ed738be))
- module type ([26fa99b](https://github.com/asset-projects/token-list/commit/26fa99b33b62fbf7b011e8950e71b54e43c2d08c))
- package.json ([b67585d](https://github.com/asset-projects/token-list/commit/b67585dfc793612212a6c4bf27b6b69a5067a13b))
- package.json ([#4](https://github.com/asset-projects/token-list/issues/4)) ([e10001c](https://github.com/asset-projects/token-list/commit/e10001c8103d3c6a34648032d674e6acb84472f6))
- rEADME ([#7](https://github.com/asset-projects/token-list/issues/7)) ([25cfe91](https://github.com/asset-projects/token-list/commit/25cfe911bcf10a0a88b0b907c39483427018af4c))
- **release.yml:** release.yml ([#6](https://github.com/asset-projects/token-list/issues/6)) ([6baacb6](https://github.com/asset-projects/token-list/commit/6baacb6bcdaaff45b908dd6a45a4b3aacf017071))


src/index.test.ts
import { tokenList } from '.';

describe('tokenList', () => {
  it('Does the data exist?', () => {
    expect(tokenList()).toBeDefined();
  });

  it('mainnet', () => {
    expect(tokenList().erc20.mainnet).toBeDefined();
  });

  it('optimistic-ethereum', () => {
    expect(tokenList().erc20['optimistic-ethereum']).toBeDefined();
  });

  it('arbitrum', () => {
    expect(tokenList().erc20.arbitrum).toBeDefined();
  });

  it('polygon', () => {
    expect(tokenList().erc20['binance-smart-chain']).toBeDefined();
  });

  // Test Network
  it('ropsten', () => {
    expect(tokenList().erc20.ropsten).toBeDefined();
  });

  it('kovan', () => {
    expect(tokenList().erc20.kovan).toBeDefined();
  });

  it('rinkeby', () => {
    expect(tokenList().erc20.rinkeby).toBeDefined();
  });

  it('goerli', () => {
    expect(tokenList().erc20.goerli).toBeDefined();
  });

  it('arbitrum-test-rinkeby', () => {
    expect(tokenList().erc20['arbitrum-test-rinkeby']).toBeDefined();
  });

  it('optimistic-ethereum-kovan', () => {
    expect(tokenList().erc20['optimistic-ethereum-kovan']).toBeDefined();
  });
});


test/index.test.ts
import { tokens, getTokenList } from '../src';

describe('Test', () => {
  it('result', () => {
    expect(getTokenList(1)).toBeDefined();
    expect(tokens()).toBeDefined();

    // Does not exist.
    expect(getTokenList(111)).toBeUndefined();
  });

  it('mainnet', () => {
    const token = getTokenList(1);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toBeDefined();
  });

  it('optimistic-ethereum', () => {
    const token = getTokenList(10);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toBeDefined();
  });

  it('arbitrum', () => {
    const token = getTokenList(42161);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toBeDefined();
  });

  it('polygon', () => {
    const token = getTokenList(137);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toBeDefined();
  });

  it('binance', () => {
    const token = getTokenList(56);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toHaveLength(0);
  });

  it('avalanche', () => {
    const token = getTokenList(43114);
    expect(token.ERC20).toBeDefined();
    expect(token.ERC721).toHaveLength(0);
  });

  // Test Network
  it('ropsten', () => {
    const token = getTokenList(3);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });

  it('kovan', () => {
    const token = getTokenList(42);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });

  it('rinkeby', () => {
    const token = getTokenList(4);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });

  it('goerli', () => {
    const token = getTokenList(5);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });

  it('arbitrum-test-rinkeby', () => {
    const token = getTokenList(421611);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });

  it('optimistic-ethereum-kovan', () => {
    const token = getTokenList(69);
    expect(token.ERC20).toHaveLength(0);
    expect(token.ERC721).toHaveLength(0);
  });
});


src/parts.ts
/** ERC20 */
import MainnetERC20Tokens from './tokens/erc20/mainnet.json';
import OptimisticEthereumERC20Tokens from './tokens/erc20/optimistic-ethereum.json';
import ArbitrumERC20Tokens from './tokens/erc20/arbitrum.json';
import PolygonERC20Tokens from './tokens/erc20/polygon.json';
import BinanceSmartChainMainnnetERC20Tokens from './tokens/erc20/binance-smart-chain-mainnet.json';
import RopstenERC20Tokens from './tokens/erc20/ropsten.json';
import KovanERC20Tokens from './tokens/erc20/ropsten.json';
import RinkebyERC20Tokens from './tokens/erc20/rinkeby.json';
import GoerliERC20Tokens from './tokens/erc20/goerli.json';
import ArbitrumTestNetworkERC20Tokens from './tokens/erc20/arbitrum-testnet-rinkeby.json';
import OptimisticEthereumKovanERC20Tokens from './tokens/erc20/optimistic-ethereum-kovan.json';
import AvalancheERC20Tokens from './tokens/erc20/avalanche.json';

/** ERC721 */
import MainnetERC721Tokens from './tokens/erc721/mainnet.json';
import OptimistiEthereumERC721Tokens from './tokens/erc721/optimistic-ethereum.json';
import ArbitrumERC721Tokens from './tokens/erc721/arbitrum.json';
import PolygonERC721Tokens from './tokens/erc721/polygon.json';

export const MAINNET = {
  ERC20: MainnetERC20Tokens,
  ERC721: MainnetERC721Tokens,
};
export const OPTIMISTIC_ETHEREUM = {
  ERC20: OptimisticEthereumERC20Tokens,
  ERC721: OptimistiEthereumERC721Tokens,
};
export const ARBITRUM = {
  ERC20: ArbitrumERC20Tokens,
  ERC721: ArbitrumERC721Tokens,
};
export const POLYGON = {
  ERC20: PolygonERC20Tokens,
  ERC721: PolygonERC721Tokens,
};
export const BINANCE_SMART_CHAIN = {
  ERC20: BinanceSmartChainMainnnetERC20Tokens,
  ERC721: [],
};
export const AVALANCHE = {
  ERC20: AvalancheERC20Tokens,
  ERC721: [],
};

export const ROPSTEN = {
  ERC20: RopstenERC20Tokens,
  ERC721: [],
};
export const KOVAN = {
  ERC20: KovanERC20Tokens,
  ERC721: [],
};
export const RINKEBY = {
  ERC20: RinkebyERC20Tokens,
  ERC721: [],
};
export const GOERLI = {
  ERC20: GoerliERC20Tokens,
  ERC721: [],
};
export const ARBITRUM_TEST_NETWORK = {
  ERC20: ArbitrumTestNetworkERC20Tokens,
  ERC721: [],
};
export const OPTIMISTIC_ETHEREUM_KOVAN = {
  ERC20: OptimisticEthereumKovanERC20Tokens,
  ERC721: [],
};


jest.config.ts
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/gl/hl9ll3wd7k7cp85z1jms76x80000gn/T/jest_dx",

  // Automatically clear mock calls, instances and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  coverageDirectory: 'coverage',

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  // coverageProvider: "babel",

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "json",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state before every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state and implementation before every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  // testEnvironment: "jest-environment-node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jest-circus/runner",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/",
  //   "\\.pnp\\.[^\\/]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


src/index.ts
export * from './types';
import type { FungibleToken, NonFungibleToken } from './types';

/** ERC20 */
import EthereumERC20Tokens from './tokens/erc20/mainnet.json';
import OptimisticEthereumERC20Tokens from './tokens/erc20/optimistic-ethereum.json';
import ArbitrumERC20Tokens from './tokens/erc20/arbitrum.json';
import PolygonERC20Tokens from './tokens/erc20/polygon.json';
import BinanceSmartChainMainnnetERC20Tokens from './tokens/erc20/binance-smart-chain-mainnet.json';
import AvalancheERC20Tokens from './tokens/erc20/avalanche.json';

/** ERC721 */
import EthereumNonFungibleTokens from './tokens/erc721/mainnet.json';
import OptimisticEthereumNonFungibleTokens from './tokens/erc721/optimistic-ethereum.json';
import ArbitrumNonFungibleTokens from './tokens/erc721/arbitrum.json';
import PolygonNonFungibleTokens from './tokens/erc721/polygon.json';

export const getTokenList = (chainId: number) => {
  switch (chainId) {
    case 1:
      return {
        ERC20: EthereumERC20Tokens,
        ERC721: EthereumNonFungibleTokens,
      };
    case 3:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 4:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 5:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 42:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 10:
      return {
        ERC20: OptimisticEthereumERC20Tokens,
        ERC721: OptimisticEthereumNonFungibleTokens,
      };
    case 69:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 42161:
      return {
        ERC20: ArbitrumERC20Tokens,
        ERC721: ArbitrumNonFungibleTokens,
      };
    case 421611:
      return {
        ERC20: [],
        ERC721: [],
      };
    case 137:
      return {
        ERC20: PolygonERC20Tokens,
        ERC721: PolygonNonFungibleTokens,
      };
    case 56:
      return {
        ERC20: BinanceSmartChainMainnnetERC20Tokens,
        ERC721: [],
      };
    case 43114:
      return {
        ERC20: AvalancheERC20Tokens,
        ERC721: [],
      };
    default:
      return undefined;
  }
};

export const tokens = (): (FungibleToken | NonFungibleToken)[] => {
  return [
    ...EthereumERC20Tokens,
    ...EthereumNonFungibleTokens,
    ...OptimisticEthereumERC20Tokens,
    ...OptimisticEthereumNonFungibleTokens,
    ...ArbitrumERC20Tokens,
    ...ArbitrumNonFungibleTokens,
    ...PolygonERC20Tokens,
    ...PolygonNonFungibleTokens,
    ...BinanceSmartChainMainnnetERC20Tokens,
    ...AvalancheERC20Tokens,
  ] as (FungibleToken | NonFungibleToken)[];
};
